distributed/core.py

```diff
@@ -495,7 +638,12 @@ async def listen(self, port_or_addr=None, allow_offload=True, **kwargs):
         )
         self.listeners.append(listener)
 
-    async def handle_comm(self, comm):
+    def handle_comm(self, comm):
+        """Start a background task that dispatches new communications to coroutine-handlers"""
+        self._ongoing_background_tasks.call_soon(self._handle_comm, comm)
+        return NoOpAwaitable()
```

Comment:
I think this might help with new connections that are accepted after the worker starts closing

```suggestion
    def handle_comm(self, comm):
        """Start a background task that dispatches new communications to coroutine-handlers"""
        try:
            self._ongoing_background_tasks.call_soon(self._handle_comm, comm)
        except AsyncTaskGroupClosedError:
            comm.abort()
        return NoOpAwaitable()
```

Reply:
which is the same as on `main`:

```
 graingert@superjacent  ~/projects/distributed   main  pytest distributed/tests/test_scheduler.py::test_missing_data_errant_worker -s
================================================================================================================================== test session starts ===================================================================================================================================
platform linux -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0 -- /home/graingert/miniconda3/envs/dask-distributed/bin/python
cachedir: .pytest_cache
rootdir: /home/graingert/projects/distributed, configfile: setup.cfg
plugins: cov-3.0.0, repeat-0.9.1, rerunfailures-10.2, timeout-2.1.0
timeout: 300.0s
timeout method: thread
timeout func_only: False
collected 1 item                                                                                                                                                                                                                                                                         

distributed/tests/test_scheduler.py::test_missing_data_errant_worker 2022-06-08 17:32:30,332 - distributed.http.proxy - INFO - To route to workers diagnostics web server please install jupyter-server-proxy: python -m pip install jupyter-server-proxy
2022-06-08 17:32:30,336 - distributed.scheduler - INFO - State start
2022-06-08 17:32:30,342 - distributed.scheduler - INFO - Clear task state
2022-06-08 17:32:30,343 - distributed.scheduler - INFO -   Scheduler at:     tcp://127.0.0.1:33183
2022-06-08 17:32:30,343 - distributed.scheduler - INFO -   dashboard at:           127.0.0.1:32993
2022-06-08 17:32:30,361 - distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:43121
2022-06-08 17:32:30,361 - distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:43121
2022-06-08 17:32:30,361 - distributed.worker - INFO -          dashboard at:            127.0.0.1:45323
2022-06-08 17:32:30,361 - distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:33183
2022-06-08 17:32:30,362 - distributed.worker - INFO - -------------------------------------------------
2022-06-08 17:32:30,362 - distributed.worker - INFO -               Threads:                          1
2022-06-08 17:32:30,362 - distributed.worker - INFO -                Memory:                  15.36 GiB
2022-06-08 17:32:30,362 - distributed.worker - INFO -       Local Directory: /tmp/tmprhlds1rj/dask-worker-space/worker-qprmb83h
2022-06-08 17:32:30,362 - distributed.worker - INFO - -------------------------------------------------
2022-06-08 17:32:30,363 - distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:39179
2022-06-08 17:32:30,363 - distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:39179
2022-06-08 17:32:30,363 - distributed.worker - INFO -          dashboard at:            127.0.0.1:46715
2022-06-08 17:32:30,363 - distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:33183
2022-06-08 17:32:30,363 - distributed.worker - INFO - -------------------------------------------------
2022-06-08 17:32:30,363 - distributed.worker - INFO -               Threads:                          1
2022-06-08 17:32:30,364 - distributed.worker - INFO -                Memory:                  15.36 GiB
2022-06-08 17:32:30,364 - distributed.worker - INFO -       Local Directory: /tmp/tmprhlds1rj/dask-worker-space/worker-2dmt431u
2022-06-08 17:32:30,364 - distributed.worker - INFO - -------------------------------------------------
2022-06-08 17:32:30,367 - distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:36953
2022-06-08 17:32:30,367 - distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:36953
2022-06-08 17:32:30,367 - distributed.worker - INFO -          dashboard at:            127.0.0.1:37997
2022-06-08 17:32:30,367 - distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:33183
2022-06-08 17:32:30,367 - distributed.worker - INFO - -------------------------------------------------
2022-06-08 17:32:30,367 - distributed.worker - INFO -               Threads:                          1
2022-06-08 17:32:30,367 - distributed.worker - INFO -                Memory:                  15.36 GiB
2022-06-08 17:32:30,367 - distributed.worker - INFO -       Local Directory: /tmp/tmprhlds1rj/dask-worker-space/worker-5vvbjqd7
2022-06-08 17:32:30,367 - distributed.worker - INFO - -------------------------------------------------
2022-06-08 17:32:31,052 - distributed.scheduler - INFO - Register worker <WorkerState 'tcp://127.0.0.1:43121', name: 0, status: init, memory: 0, processing: 0>
2022-06-08 17:32:31,053 - distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:43121
2022-06-08 17:32:31,054 - distributed.core - INFO - Starting established connection
2022-06-08 17:32:31,054 - distributed.scheduler - INFO - Register worker <WorkerState 'tcp://127.0.0.1:39179', name: 1, status: init, memory: 0, processing: 0>
2022-06-08 17:32:31,055 - distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:39179
2022-06-08 17:32:31,055 - distributed.core - INFO - Starting established connection
2022-06-08 17:32:31,056 - distributed.scheduler - INFO - Register worker <WorkerState 'tcp://127.0.0.1:36953', name: 2, status: init, memory: 0, processing: 0>
2022-06-08 17:32:31,057 - distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:36953
2022-06-08 17:32:31,057 - distributed.core - INFO - Starting established connection
2022-06-08 17:32:31,057 - distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:33183
2022-06-08 17:32:31,057 - distributed.worker - INFO - -------------------------------------------------
2022-06-08 17:32:31,058 - distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:33183
2022-06-08 17:32:31,058 - distributed.worker - INFO - -------------------------------------------------
2022-06-08 17:32:31,059 - distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:33183
2022-06-08 17:32:31,059 - distributed.worker - INFO - -------------------------------------------------
2022-06-08 17:32:31,059 - distributed.core - INFO - Starting established connection
2022-06-08 17:32:31,060 - distributed.core - INFO - Starting established connection
2022-06-08 17:32:31,060 - distributed.core - INFO - Starting established connection
2022-06-08 17:32:31,098 - distributed.scheduler - INFO - Receive client connection: Client-97ba5f2e-e748-11ec-96e8-9cb6d08f15e1
2022-06-08 17:32:31,099 - distributed.core - INFO - Starting established connection
2022-06-08 17:32:31,587 - distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:43121
2022-06-08 17:32:31,605 - distributed.scheduler - INFO - Remove worker <WorkerState 'tcp://127.0.0.1:43121', name: 0, status: closing, memory: 1, processing: 0>
2022-06-08 17:32:31,605 - distributed.core - INFO - Removing comms to tcp://127.0.0.1:43121
2022-06-08 17:32:31,606 - distributed.worker - INFO - Connection to scheduler broken. Closing without reporting. ID: Worker-b33593a2-312a-4bcb-a4aa-8e33d80554c8 Address tcp://127.0.0.1:43121 Status: Status.closing
2022-06-08 17:32:31,609 - distributed.worker - ERROR - Worker stream died during communication: tcp://127.0.0.1:43121
Traceback (most recent call last):
  File "/home/graingert/miniconda3/envs/dask-distributed/lib/python3.10/site-packages/tornado/iostream.py", line 867, in _read_to_buffer
    bytes_read = self.read_from_fd(buf)
  File "/home/graingert/miniconda3/envs/dask-distributed/lib/python3.10/site-packages/tornado/iostream.py", line 1140, in read_from_fd
    return self.socket.recv_into(buf, len(buf))
ConnectionResetError: [Errno 104] Connection reset by peer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/graingert/projects/distributed/distributed/worker.py", line 3290, in gather_dep
    response = await get_data_from_worker(
  File "/home/graingert/projects/distributed/distributed/worker.py", line 4611, in get_data_from_worker
    return await retry_operation(_get_data, operation="get_data_from_worker")
  File "/home/graingert/projects/distributed/distributed/utils_comm.py", line 381, in retry_operation
    return await retry(
  File "/home/graingert/projects/distributed/distributed/utils_comm.py", line 366, in retry
    return await coro()
  File "/home/graingert/projects/distributed/distributed/worker.py", line 4591, in _get_data
    response = await send_recv(
  File "/home/graingert/projects/distributed/distributed/core.py", line 748, in send_recv
    response = await comm.read(deserializers=deserializers)
  File "/home/graingert/projects/distributed/distributed/comm/tcp.py", line 242, in read
    convert_stream_closed_error(self, e)
  File "/home/graingert/projects/distributed/distributed/comm/tcp.py", line 148, in convert_stream_closed_error
    raise CommClosedError(f"in {obj}: {exc.__class__.__name__}: {exc}") from exc
distributed.comm.core.CommClosedError: in <TCP (closed) Ephemeral Worker->Worker for gather local=tcp://127.0.0.1:53490 remote=tcp://127.0.0.1:43121>: ConnectionResetError: [Errno 104] Connection reset by peer
2022-06-08 17:32:31,775 - distributed.scheduler - INFO - Remove client Client-97ba5f2e-e748-11ec-96e8-9cb6d08f15e1
2022-06-08 17:32:31,776 - distributed.scheduler - INFO - Remove client Client-97ba5f2e-e748-11ec-96e8-9cb6d08f15e1
2022-06-08 17:32:31,776 - distributed.scheduler - INFO - Close client connection: Client-97ba5f2e-e748-11ec-96e8-9cb6d08f15e1
2022-06-08 17:32:31,777 - distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:39179
2022-06-08 17:32:31,778 - distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:36953
2022-06-08 17:32:31,780 - distributed.scheduler - INFO - Remove worker <WorkerState 'tcp://127.0.0.1:39179', name: 1, status: closing, memory: 0, processing: 0>
2022-06-08 17:32:31,780 - distributed.core - INFO - Removing comms to tcp://127.0.0.1:39179
2022-06-08 17:32:31,781 - distributed.scheduler - INFO - Remove worker <WorkerState 'tcp://127.0.0.1:36953', name: 2, status: closing, memory: 0, processing: 0>
2022-06-08 17:32:31,781 - distributed.core - INFO - Removing comms to tcp://127.0.0.1:36953
2022-06-08 17:32:31,781 - distributed.scheduler - INFO - Lost all workers
2022-06-08 17:32:31,781 - distributed.worker - INFO - Connection to scheduler broken. Closing without reporting. ID: Worker-600a21d1-a19a-4f79-9db6-322ea9c286bb Address tcp://127.0.0.1:39179 Status: Status.closing
2022-06-08 17:32:31,781 - distributed.worker - INFO - Connection to scheduler broken. Closing without reporting. ID: Worker-5d99f3f3-c0ed-4531-b2c0-e271e50746b4 Address tcp://127.0.0.1:36953 Status: Status.closing
2022-06-08 17:32:31,789 - distributed.scheduler - INFO - Scheduler closing...
2022-06-08 17:32:31,789 - distributed.scheduler - INFO - Scheduler closing all comms
PASSED

================================================================================================================================== slowest 20 durations ==================================================================================================================================
1.55s call     distributed/tests/test_scheduler.py::test_missing_data_errant_worker

(2 durations < 0.005s hidden.  Use -vv to show these durations.)
=================================================================================================================================== 1 passed in 1.73s ====================================================================================================================================
```

Reply:
Done